precompiles/bank/method_withdraw.go

```diff
@@ -0,0 +1,179 @@
+	evm *vm.EVM,
+	contract *vm.Contract,
+	method *abi.Method,
+	args []interface{},
+) (result []byte, err error) {
+	// 1. Check everything is correct.
+	if len(args) != 2 {
+		return nil, &(ptypes.ErrInvalidNumberOfArgs{
+			Got:    len(args),
+			Expect: 2,
+		})
+	}
+
+	// Unpack parameters for function withdraw.
+	// function withdraw(address zrc20, uint256 amount) external returns (bool success);
+	zrc20Addr, amount, err := unpackWithdrawArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the correct caller address.
+	caller, err := getEVMCallerAddress(evm, contract)
+	if err != nil {
+		return nil, err
+	}
+
+	// Get the cosmos address of the caller.
+	// This address should have enough cosmos coin balance as the requested amount.
+	fromAddr, err := getCosmosAddress(c.bankKeeper, caller)
+	if err != nil {
+		return nil, err
+	}
+
+	// Safety check: token has to be a valid whitelisted ZRC20 and not be paused.
+	t, found := c.fungibleKeeper.GetForeignCoins(ctx, zrc20Addr.String())
+	if !found || t.Paused {
+		return nil, &ptypes.ErrInvalidToken{
+			Got:    zrc20Addr.String(),
+			Reason: "token is not a whitelisted ZRC20 or it's paused",
+		}
+	}
+
+	// Caller has to have enough cosmos coin balance to withdraw the requested amount.
+	coin := c.bankKeeper.GetBalance(ctx, fromAddr, ZRC20ToCosmosDenom(zrc20Addr))
+	if coin.Amount.IsNil() {
+		return nil, &ptypes.ErrInsufficientBalance{
+			Requested: amount.String(),
+			Got:       "nil",
+		}
+	}
+
+	if coin.Amount.LT(math.NewIntFromBigInt(amount)) {
+		return nil, &ptypes.ErrInsufficientBalance{
+			Requested: amount.String(),
+			Got:       coin.Amount.String(),
+		}
+	}
+
+	coinSet, err := createCoinSet(ZRC20ToCosmosDenom(zrc20Addr), amount)
+	if err != nil {
+		return nil, err
+	}
+
+	// Check for bank's ZRC20 balance.
+	// function balanceOf(address account) public view virtual override returns (uint256)
+	resBalanceOf, err := c.CallContract(
+		ctx,
+		&c.fungibleKeeper,
+		c.zrc20ABI,
+		zrc20Addr,
+		"balanceOf",
+		[]interface{}{ContractAddress},
+	)
+	if err != nil {
+		return nil, &ptypes.ErrUnexpected{
+			When: "balanceOf",
+			Got:  err.Error(),
+		}
+	}
+
+	balance, ok := resBalanceOf[0].(*big.Int)
+	if !ok || balance.Cmp(amount) == -1 {
+		return nil, &ptypes.ErrInvalidAmount{
+			Got: "not enough bank balance",
+		}
+	}
+
+	// 2. Effect: transfer balance.
+
+	// function transfer(address recipient, uint256 amount) public virtual override returns (bool)
+	resTransferFrom, err := c.CallContract(
+		ctx,
+		&c.fungibleKeeper,
+		c.zrc20ABI,
+		zrc20Addr,
+		"transfer",
+		[]interface{}{caller /* sender */, amount},
+	)
+	if err != nil {
+		return nil, &ptypes.ErrUnexpected{
+			When: "transferFrom",
+			Got:  err.Error(),
+		}
+	}
+
+	transferred, ok := resTransferFrom[0].(bool)
+	if !ok || !transferred {
+		return nil, &ptypes.ErrUnexpected{
+			When: "transferFrom",
+			Got:  "transaction not successful",
+		}
+	}
+
+	// 3. Interactions: send to module and burn.
```

Comment:
Even if the call is atomic, as a security measure, we should first do this operation before transferring the ZRC20 back

Reply:
Modified in af858abd8aaa8c7087e6ec75abe7299e34279220